plugins/faustwp/includes/replacement/callbacks.php

```diff
@@ -31,36 +31,42 @@
  *
  * @return string The post content.
  */
-function content_replacement( $content ) {
-	$use_wp_domain_for_permalinks = ! domain_replacement_enabled();
-	$use_wp_domain_for_media      = use_wp_domain_for_media();
+function content_replacement( string $content ): string {
 
-	if ( $use_wp_domain_for_permalinks && $use_wp_domain_for_media ) {
+	if ( ! $content ) {
+		return '';
+	}
+
+	$replace_content_urls = domain_replacement_enabled();
+	$replace_media_urls   = ! use_wp_domain_for_media();
+
+	if ( ! $replace_content_urls && ! $replace_media_urls ) {
 		return $content;
 	}
 
-	$replacement = faustwp_get_setting( 'frontend_uri' );
-	if ( ! $replacement ) {
-		$replacement = '/';
+	$wp_site_urls = faustwp_get_wp_site_urls();
+	if ( empty( $wp_site_urls ) ) {
+		return $content;
 	}
 
-	$site_url  = site_url();
-	$media_dir = str_replace( $site_url, '', wp_upload_dir()['baseurl'] );
-	$media_url = $site_url . $media_dir;
+	$wp_media_urls       = faustwp_get_wp_media_urls();
+	$relative_upload_url = faustwp_get_relative_upload_url( $wp_site_urls );
+	$frontend_uri        = faustwp_get_setting( 'frontend_uri' );
+	if ( ! $frontend_uri ) {
+		$frontend_uri = '/';
+	}
 
-	if ( $use_wp_domain_for_permalinks && ! $use_wp_domain_for_media ) {
-		$content = str_replace( $media_url, $replacement . $media_dir, $content );
-		return $content;
+	if ( $replace_content_urls && $replace_media_urls ) {
+		return str_replace( $wp_site_urls, $frontend_uri, $content );
 	}
 
-	if ( ! $use_wp_domain_for_permalinks && ! $use_wp_domain_for_media ) {
-		$content = str_replace( $site_url, $replacement, $content );
-		return $content;
+	if ( $replace_media_urls ) {
+		return faustwp_replace_media_url( $content, $wp_media_urls, $frontend_uri . $relative_upload_url );
 	}
 
-	if ( ! $use_wp_domain_for_permalinks && $use_wp_domain_for_media ) {
-		$content = preg_replace( "#{$site_url}(?!{$media_dir})#", "{$replacement}", $content );
-		return $content;
+	foreach ( $wp_site_urls as $site_url ) {
+		$pattern_exclude_media_urls = '#' . preg_quote( $site_url, '#' ) . "(?!{$relative_upload_url}(\/|$))#";
```

Comment:
I believe that we can implode the urls into the pattern itself instead of preg_replace it multiple times inside foreach. I guess if the urls array is less then ~50 it is faster and more compact.

Reply:
I tried to use implode but the only way I was able to do this is with array map. I don't know if the performance is worth it compared to readibility as I find the array map part not very readable.

```php
	$pattern_exclude_media_urls = '#(' . implode( '|', array_map( function( $url ) {
			return preg_quote( $url, '#' );
		}, $wp_site_urls ) ) . ")(?!{$relative_upload_url}(\/|$))#";

	return preg_replace( $pattern_exclude_media_urls, $frontend_uri, $content );
```

Happy for a better suggesting on how to do the implode part 👍

Reply:
I was actually able to tidy up the array_map based on your suggestion above so thanks 😄